File: apps/desktop/src/api/llm-settings.ts
import { apiGet, apiPut } from "./client"

export type LlmProviderId = "deepseek" | "openai_compatible"

export interface LlmProviderOption {
  id: LlmProviderId
  label: string
  requires_base_url: boolean
  default_model: string
  default_base_url: string
}

export interface LlmSettings {
  provider: LlmProviderId
  model: string
  base_url: string
  providers: LlmProviderOption[]
}

export interface LlmModelOption {
  id: string
}

export interface LlmModelsResponse {
  provider: LlmProviderId
  current_model: string
  available: boolean
  models: LlmModelOption[]
  detail: string
}

export interface LlmSettingsUpdate {
  provider: LlmProviderId
  model: string
  base_url: string
}

export type LlmRuntimeState = "available" | "calling" | "unavailable"

export interface LlmRuntimeStatus {
  state: LlmRuntimeState
  provider: string
  model: string
  detail: string
  active_requests: number
}

const LLM_SETTINGS_PATH = "/api/settings/llm"

export function getLlmSettings(): Promise<LlmSettings> {
  return apiGet<LlmSettings>(LLM_SETTINGS_PATH)
}

export function getLlmRuntimeStatus(): Promise<LlmRuntimeStatus> {
  return apiGet<LlmRuntimeStatus>(`${LLM_SETTINGS_PATH}/status`)
}

export function getAvailableLlmModels(): Promise<LlmModelsResponse> {
  return apiGet<LlmModelsResponse>(`${LLM_SETTINGS_PATH}/models`)
}

export function updateLlmSettings(payload: LlmSettingsUpdate): Promise<LlmSettings> {
  return apiPut<LlmSettings, LlmSettingsUpdate>(LLM_SETTINGS_PATH, payload)
}

File: backend/models/llm_settings.py
"""Safe request and response contracts for local LLM configuration."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

AIProviderName = Literal[
    "deepseek",
    "openai",
    "google",
    "mistral",
    "groq",
    "openrouter",
    "together",
    "qwen",
    "openai_compatible",
]
class LLMProviderOption(BaseModel):
    id: AIProviderName
    label: str
    requires_base_url: bool
    default_model: str
    default_base_url: str


class LLMSettingsResponse(BaseModel):
    provider: AIProviderName
    model: str
    base_url: str
    providers: list[LLMProviderOption] = Field(default_factory=list)


class LLMModelOption(BaseModel):
    id: str = Field(min_length=1, max_length=512)


class LLMModelsResponse(BaseModel):
    provider: AIProviderName
    current_model: str = ""
    available: bool = False
    models: list[LLMModelOption] = Field(default_factory=list)
    detail: str = ""


class LLMSettingsUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: AIProviderName
    model: str = Field(min_length=1, max_length=256)
    base_url: str = Field(default="", max_length=2048)


class LLMRuntimeStatusResponse(BaseModel):
    state: Literal["available", "calling", "unavailable"]
    provider: str = ""
    model: str = ""
    detail: str = ""
    active_requests: int = Field(default=0, ge=0)


__all__ = [
    "AIProviderName",
    "LLMModelOption",
    "LLMModelsResponse",
    "LLMProviderOption",
    "LLMRuntimeStatusResponse",
    "LLMSettingsResponse",
    "LLMSettingsUpdateRequest",
]
